File: backend/app/transliterator.py
"""Uzbek Latin <-> Cyrillic Transliteration Engine.

Implements official Uzbek phonology rules, digraph precedence, apostrophe
normalization, case-preservation, and contextual 'E'/'Э' rules.
"""

import re
from typing import Literal

# Comprehensive apostrophe variants matching official standards and common OCR/keyboard symbols
# \u2018: ‘ (left single quote)
# \u2019: ’ (right single quote)
# \u201A: ‚ (single low-9 quote)
# \u201B: ‛ (single high-reversed-9 quote)
# \u02BB: ʻ (modifier letter turned comma - official Uzbek Latin okina)
# \u02BC: ʼ (modifier letter apostrophe)
# \u02BD: ʽ (modifier letter reversed comma)
# \u0060: ` (grave accent)
# \u00B4: ´ (acute accent)
# \u02B9: ʹ (modifier letter prime)
# \u02BA: ʺ (modifier letter double prime)
# \u2032: ′ (prime)
# \u2033: ″ (double prime)
APOSTROPHE_VARIANTS = re.compile(
    r"[\u2018\u2019\u201A\u201B\u02BB\u02BC\u02BD\u0060\u00B4\u02B9\u02BA\u2032\u2033]"
)

# Vowels in Latin and Cyrillic (for contextual 'E'/'Э' rules)
LATIN_VOWELS = set("AEIOUaeiou")
CYRILLIC_VOWELS = set("АЕЁИОУЭЮЯЎаеёиоуэюяў")

# Punctuation/delimiters that delineate word boundaries
WORD_BOUNDARY_CHARS = set(" \t\n\r\f\v.,!?:;()[]{}<>/\\\"'«»—–-_#@$%^&*+=~`|")


def normalize_apostrophes(text: str) -> str:
    """Normalizes all apostrophe variants to standard typewriter apostrophe ' (U+0027)."""
    return APOSTROPHE_VARIANTS.sub("'", text)


def is_uppercase_context(text: str, index: int) -> bool:
    """Determines if a letter at `index` is in an ALL_CAPS context (e.g. SHAHAR vs Shahar)."""
    # Check next char
    if index + 1 < len(text) and text[index + 1].isupper():
        return True
    # Check previous char
    if index - 1 >= 0 and text[index - 1].isupper():
        return True
    return False


def latin_to_cyrillic(text: str) -> str:
    """Converts Uzbek text from Latin to Cyrillic script.

    Adheres to:
    - Apostrophe normalization.
    - Compound sound protection: s'h -> сҳ (not ш).
    - Digraph precedence: yo', yu', sh, ch, yo, yu, ya, ye, o', g'.
    - Initial/post-vocalic 'e' -> 'э'; post-consonantal 'e' -> 'е'.
    - Complete Uzbek alphabet mapping.
    - Trailing/stray apostrophe -> 'ъ'.
    """
    if not text:
        return ""

    text = normalize_apostrophes(text)
    length = len(text)
    result = []
    i = 0

    while i < length:
        ch = text[i]
        next_ch = text[i + 1] if i + 1 < length else ""
        next_next_ch = text[i + 2] if i + 2 < length else ""

        # 1. Non-digraph compound: s'h / S'h / S'H / s'H -> сҳ / Сҳ / СҲ / сҲ
        if ch in ("s", "S") and next_ch == "'" and next_next_ch in ("h", "H"):
            if ch == "S" and next_next_ch == "H":
                result.append("СҲ")
            elif ch == "S":
                result.append("Сҳ")
            elif next_next_ch == "H":
                result.append("сҲ")
            else:
                result.append("сҳ")
            i += 3
            continue

        # 2. Triple letter digraph with apostrophe: yo' / Yo' / YO' -> йў / Йў / ЙЎ
        # In Uzbek: "yo'l" -> "йўл", NOT "ёъл"
        if ch in ("y", "Y") and next_ch in ("o", "O") and next_next_ch == "'":
            if ch == "Y" and next_ch == "O":
                result.append("ЙЎ")
            elif ch == "Y":
                result.append("Йў")
            elif next_ch == "O":
                result.append("йЎ")
            else:
                result.append("йў")
            i += 3
            continue

        # 3. Digraphs with apostrophes: O' / o', G' / g'
        if next_ch == "'":
            if ch == "O":
                result.append("Ў")
                i += 2
                continue
            elif ch == "o":
                result.append("ў")
                i += 2
                continue
            elif ch == "G":
                result.append("Ғ")
                i += 2
                continue
            elif ch == "g":
                result.append("ғ")
                i += 2
                continue

        # 4. Standard Two-letter Digraphs: Sh, Ch, Yo, Yu, Ya, Ye
        pair = ch + next_ch
        pair_lower = pair.lower()

        if pair_lower in ("sh", "ch", "yo", "yu", "ya", "ye"):
            is_all_caps = ch.isupper() and next_ch.isupper()
            is_title = ch.isupper() and not next_ch.isupper()

            digraph_map = {
                "sh": ("Ш" if is_all_caps else ("Ш" if is_title else "ш")),
                "ch": ("Ч" if is_all_caps else ("Ч" if is_title else "ч")),
                "yo": ("Ё" if is_all_caps else ("Ё" if is_title else "ё")),
                "yu": ("Ю" if is_all_caps else ("Ю" if is_title else "ю")),
                "ya": ("Я" if is_all_caps else ("Я" if is_title else "я")),
                "ye": ("Е" if is_all_caps else ("Е" if is_title else "е")),
            }
            result.append(digraph_map[pair_lower])
            i += 2
            continue

        # 5. Loanword Digraph: ts / Ts / TS -> ц / Ц
        # Matches word-start (tsirk -> цирк) or loanword suffixes (konstitutsiya -> конституция, litsey -> лицей)
        # Prevents breaking conditional verbs ending in -tsa (ketsa -> кетса, aytsa -> айтса)
        if ch in ("t", "T") and next_ch in ("s", "S"):
            prev_ch = text[i - 1] if i > 0 else ""
            is_word_start = i == 0 or prev_ch in WORD_BOUNDARY_CHARS
            is_followed_by_vowel = next_next_ch.lower() in ("i", "e", "o", "u", "y")

            if is_word_start or is_followed_by_vowel:
                is_all_caps = ch.isupper() and next_ch.isupper()
                is_title = ch.isupper() and not next_ch.isupper()
                result.append("Ц" if (is_all_caps or is_title) else "ц")
                i += 2
                continue

        # 5. Contextual 'E' / 'e' rule
        if ch in ("E", "e"):
            prev_ch = text[i - 1] if i > 0 else ""
            is_at_start = i == 0 or prev_ch in WORD_BOUNDARY_CHARS
            is_after_vowel = prev_ch in LATIN_VOWELS or prev_ch in CYRILLIC_VOWELS

            if is_at_start or is_after_vowel:
                result.append("Э" if ch == "E" else "э")
            else:
                result.append("Е" if ch == "E" else "е")
            i += 1
            continue

        # 6. Single letters
        single_map = {
            "A": "А", "a": "а",
            "B": "Б", "b": "б",
            "D": "Д", "d": "д",
            "F": "Ф", "f": "ф",
            "G": "Г", "g": "г",
            "H": "Ҳ", "h": "ҳ",
            "I": "И", "i": "и",
            "J": "Ж", "j": "ж",
            "K": "К", "k": "к",
            "L": "Л", "l": "л",
            "M": "М", "m": "м",
            "N": "Н", "n": "н",
            "O": "О", "o": "о",
            "P": "П", "p": "п",
            "Q": "Қ", "q": "қ",
            "R": "Р", "r": "р",
            "S": "С", "s": "с",
            "T": "Т", "t": "т",
            "U": "У", "u": "у",
            "V": "В", "v": "в",
            "X": "Х", "x": "х",
            "Y": "Й", "y": "й",
            "Z": "З", "z": "з",
            # Additional loan letters
            "C": "С", "c": "с",
            "W": "В", "w": "в",
            "'": "ъ",
        }

        if ch in single_map:
            result.append(single_map[ch])
        else:
            result.append(ch)
        i += 1

    return "".join(result)


def cyrillic_to_latin(text: str) -> str:
    """Converts Uzbek text from Cyrillic to Latin script.

    Adheres to:
    - Contextual 'Е'/'е': word-start or after vowel -> 'Ye'/'ye'; after consonant -> 'E'/'e'.
    - 'Э'/'э' -> 'E'/'e'.
    - Digraph case preservation (e.g. ШАҲАР -> SHAHAR, Шаҳар -> Shahar).
    - Compound protection: сҳ -> s'h (to distinguish from sh -> ш).
    - 'Ц'/'ц' -> 'Ts'/'ts'.
    - Tutuq belgisi 'ъ' -> typewriter apostrophe '.
    - Soft sign 'ь' -> omitted in modern Uzbek orthography.
    - Full 29-letter alphabet mapping + loan letters.
    """
    if not text:
        return ""

    length = len(text)
    result = []
    i = 0

    while i < length:
        ch = text[i]
        next_ch = text[i + 1] if i + 1 < length else ""

        # 1. Non-digraph compound: сҳ / Сҳ / СҲ / сҲ -> s'h / S'h / S'H / s'H
        if ch in ("с", "С") and next_ch in ("ҳ", "Ҳ"):
            if ch == "С" and next_ch == "Ҳ":
                result.append("S'H")
            elif ch == "С":
                result.append("S'h")
            elif next_ch == "Ҳ":
                result.append("s'H")
            else:
                result.append("s'h")
            i += 2
            continue

        # 2. Contextual 'Е' / 'е'
        if ch in ("Е", "е"):
            prev_ch = text[i - 1] if i > 0 else ""
            is_at_start = i == 0 or prev_ch in WORD_BOUNDARY_CHARS
            is_after_vowel = prev_ch in CYRILLIC_VOWELS or prev_ch in LATIN_VOWELS

            if is_at_start or is_after_vowel:
                if ch == "Е":
                    result.append("YE" if is_uppercase_context(text, i) else "Ye")
                else:
                    result.append("ye")
            else:
                result.append("E" if ch == "Е" else "e")
            i += 1
            continue

        # 3. 'Э' / 'э'
        if ch == "Э":
            result.append("E")
            i += 1
            continue
        elif ch == "э":
            result.append("e")
            i += 1
            continue

        # 4. Multi-letter representations: Ш, Ч, Ё, Ю, Я, Ц, Ў, Ғ, Щ
        upper_ctx = is_uppercase_context(text, i)

        cyr_multi_map = {
            "Ш": "SH" if upper_ctx else "Sh",
            "ш": "sh",
            "Ч": "CH" if upper_ctx else "Ch",
            "ч": "ch",
            "Ё": "YO" if upper_ctx else "Yo",
            "ё": "yo",
            "Ю": "YU" if upper_ctx else "Yu",
            "ю": "yu",
            "Я": "YA" if upper_ctx else "Ya",
            "я": "ya",
            "Ц": "TS" if upper_ctx else "Ts",
            "ц": "ts",
            "Ў": "O'",
            "ў": "o'",
            "Ғ": "G'",
            "ғ": "g'",
            "Щ": "SHCH" if upper_ctx else "Shch",
            "щ": "shch",
            "Ъ": "'",
            "ъ": "'",
            "Ь": "",
            "ь": "",
        }

        if ch in cyr_multi_map:
            result.append(cyr_multi_map[ch])
            i += 1
            continue

        # 5. Single Cyrillic letters (note: explicitly Cyrillic characters here)
        single_cyr_map = {
            "А": "A", "а": "a",
            "Б": "B", "б": "b",
            "В": "V", "в": "v",
            "Г": "G", "г": "g",
            "Д": "D", "д": "d",
            "Ж": "J", "ж": "j",
            "З": "Z", "з": "z",
            "И": "I", "и": "i",
            "Й": "Y", "й": "y",
            "К": "K", "к": "k",
            "Қ": "Q", "қ": "q",
            "Л": "L", "л": "l",
            "М": "M", "м": "m",
            "Н": "N", "н": "n",
            "О": "O", "о": "o",
            "П": "P", "п": "p",
            "Р": "R", "р": "r",
            "С": "S", "с": "s",
            "Т": "T", "т": "t",
            "У": "U", "у": "u",
            "Ф": "F", "ф": "f",
            "Х": "X", "х": "x",
            "Ҳ": "H", "ҳ": "h",
            "Ы": "I", "ы": "i",
        }

        if ch in single_cyr_map:
            result.append(single_cyr_map[ch])
        else:
            result.append(ch)
        i += 1

    return "".join(result)


def transliterate(text: str, direction: Literal["latin-to-cyrillic", "cyrillic-to-latin"] = "latin-to-cyrillic") -> str:
    """Translates text in the specified direction."""
    if direction == "latin-to-cyrillic":
        return latin_to_cyrillic(text)
    elif direction == "cyrillic-to-latin":
        return cyrillic_to_latin(text)
    else:
        raise ValueError(f"Unknown transliteration direction: {direction}")

File: frontend/js/transliterator.js
/**
 * Uzbek Latin <-> Cyrillic Transliteration Engine (JavaScript / Browser & Node.js).
 *
 * Implements official Uzbek phonology rules, digraph precedence, apostrophe
 * normalization, case-preservation, and contextual 'E'/'Э' rules.
 */

(function (root, factory) {
  if (typeof define === 'function' && define.amd) {
    define([], factory);
  } else if (typeof module === 'object' && module.exports) {
    module.exports = factory();
  } else {
    root.UzbekTransliterator = factory();
  }
}(typeof self !== 'undefined' ? self : this, function () {
  'use strict';

  // Unicode apostrophe variants normalized to standard typewriter apostrophe ' (U+0027)
  var APOSTROPHE_VARIANTS = /[\u2018\u2019\u201A\u201B\u02BB\u02BC\u02BD\u0060\u00B4\u02B9\u02BA\u2032\u2033]/g;

  var LATIN_VOWELS = new Set(['A', 'E', 'I', 'O', 'U', 'a', 'e', 'i', 'o', 'u']);
  var CYRILLIC_VOWELS = new Set(['А', 'Е', 'Ё', 'И', 'О', 'У', 'Э', 'Ю', 'Я', 'Ў', 'а', 'е', 'ё', 'и', 'о', 'у', 'э', 'ю', 'я', 'ў']);
  var WORD_BOUNDARY_CHARS = new Set(" \t\n\r\f\v.,!?:;()[]{}<>/\\\"'«»—–-_#@$%^&*+=~`|".split(''));

  function normalizeApostrophes(text) {
    if (!text) return '';
    return text.replace(APOSTROPHE_VARIANTS, "'");
  }

  function isUpper(char) {
    return char >= 'A' && char <= 'Z' || char >= 'А' && char <= 'Я' || char === 'Ў' || char === 'Ғ' || char === 'Ё';
  }

  function isUpperCaseContext(text, index) {
    if (index + 1 < text.length && isUpper(text[index + 1])) return true;
    if (index - 1 >= 0 && isUpper(text[index - 1])) return true;
    return false;
  }

  function latinToCyrillic(text) {
    if (!text) return '';
    text = normalizeApostrophes(text);
    var len = text.length;
    var result = [];
    var i = 0;

    while (i < len) {
      var ch = text[i];
      var next_ch = i + 1 < len ? text[i + 1] : '';
      var next_next_ch = i + 2 < len ? text[i + 2] : '';

      // 1. Non-digraph compound: s'h -> сҳ
      if ((ch === 's' || ch === 'S') && next_ch === "'" && (next_next_ch === 'h' || next_next_ch === 'H')) {
        if (ch === 'S' && next_next_ch === 'H') result.push('СҲ');
        else if (ch === 'S') result.push('Сҳ');
        else if (next_next_ch === 'H') result.push('сҲ');
        else result.push('сҳ');
        i += 3;
        continue;
      }

      // 2. Triple letter digraph: yo' -> йў
      if ((ch === 'y' || ch === 'Y') && (next_ch === 'o' || next_ch === 'O') && next_next_ch === "'") {
        if (ch === 'Y' && next_ch === 'O') result.push('ЙЎ');
        else if (ch === 'Y') result.push('Йў');
        else if (next_ch === 'O') result.push('йЎ');
        else result.push('йў');
        i += 3;
        continue;
      }

      // 3. Digraphs with apostrophes: O' / o', G' / g'
      if (next_ch === "'") {
        if (ch === 'O') { result.push('Ў'); i += 2; continue; }
        if (ch === 'o') { result.push('ў'); i += 2; continue; }
        if (ch === 'G') { result.push('Ғ'); i += 2; continue; }
        if (ch === 'g') { result.push('ғ'); i += 2; continue; }
      }

      // 4. Standard Two-letter Digraphs: Sh, Ch, Yo, Yu, Ya, Ye
      var pair = ch + next_ch;
      var pair_lower = pair.toLowerCase();

      if (pair_lower === 'sh' || pair_lower === 'ch' || pair_lower === 'yo' ||
          pair_lower === 'yu' || pair_lower === 'ya' || pair_lower === 'ye') {
        var is_all_caps = (ch >= 'A' && ch <= 'Z') && (next_ch >= 'A' && next_ch <= 'Z');
        var is_title = (ch >= 'A' && ch <= 'Z') && !(next_ch >= 'A' && next_ch <= 'Z');

        var digraph_map = {
          'sh': is_all_caps ? 'Ш' : (is_title ? 'Ш' : 'ш'),
          'ch': is_all_caps ? 'Ч' : (is_title ? 'Ч' : 'ч'),
          'yo': is_all_caps ? 'Ё' : (is_title ? 'Ё' : 'ё'),
          'yu': is_all_caps ? 'Ю' : (is_title ? 'Ю' : 'ю'),
          'ya': is_all_caps ? 'Я' : (is_title ? 'Я' : 'я'),
          'ye': is_all_caps ? 'Е' : (is_title ? 'Е' : 'е'),
        };
        result.push(digraph_map[pair_lower]);
        i += 2;
        continue;
      }

      // 5. Loanword Digraph: ts / Ts / TS -> ц / Ц
      // Matches word-start (tsirk -> цирк) or loanword suffixes (konstitutsiya -> конституция, litsey -> лицей)
      // Prevents breaking conditional verbs ending in -tsa (ketsa -> кетса, aytsa -> айтса)
      if ((ch === 't' || ch === 'T') && (next_ch === 's' || next_ch === 'S')) {
        var prev_ch_ts = i > 0 ? text[i - 1] : '';
        var is_word_start_ts = i === 0 || WORD_BOUNDARY_CHARS.has(prev_ch_ts);
        var next_vowel = next_next_ch.toLowerCase();
        var is_followed_by_vowel_ts = next_vowel === 'i' || next_vowel === 'e' || next_vowel === 'o' || next_vowel === 'u' || next_vowel === 'y';

        if (is_word_start_ts || is_followed_by_vowel_ts) {
          var is_all_caps_ts = (ch >= 'A' && ch <= 'Z') && (next_ch >= 'A' && next_ch <= 'Z');
          var is_title_ts = (ch >= 'A' && ch <= 'Z') && !(next_ch >= 'A' && next_ch <= 'Z');
          result.push((is_all_caps_ts || is_title_ts) ? 'Ц' : 'ц');
          i += 2;
          continue;
        }
      }

      // 5. Contextual 'E' / 'e'
      if (ch === 'E' || ch === 'e') {
        var prev_ch = i > 0 ? text[i - 1] : '';
        var is_at_start = i === 0 || WORD_BOUNDARY_CHARS.has(prev_ch);
        var is_after_vowel = LATIN_VOWELS.has(prev_ch) || CYRILLIC_VOWELS.has(prev_ch);

        if (is_at_start || is_after_vowel) {
          result.push(ch === 'E' ? 'Э' : 'э');
        } else {
          result.push(ch === 'E' ? 'Е' : 'е');
        }
        i += 1;
        continue;
      }

      // 6. Single letters
      var single_map = {
        'A': 'А', 'a': 'а',
        'B': 'Б', 'b': 'б',
        'D': 'Д', 'd': 'д',
        'F': 'Ф', 'f': 'ф',
        'G': 'Г', 'g': 'г',
        'H': 'Ҳ', 'h': 'ҳ',
        'I': 'И', 'i': 'и',
        'J': 'Ж', 'j': 'ж',
        'K': 'К', 'k': 'к',
        'L': 'Л', 'l': 'л',
        'M': 'М', 'm': 'м',
        'N': 'Н', 'n': 'н',
        'O': 'О', 'o': 'о',
        'P': 'П', 'p': 'п',
        'Q': 'Қ', 'q': 'қ',
        'R': 'Р', 'r': 'р',
        'S': 'С', 's': 'с',
        'T': 'Т', 't': 'т',
        'U': 'У', 'u': 'у',
        'V': 'В', 'v': 'в',
        'X': 'Х', 'x': 'х',
        'Y': 'Й', 'y': 'й',
        'Z': 'З', 'z': 'з',
        'C': 'С', 'c': 'с',
        'W': 'В', 'w': 'в',
        "'": 'ъ',
      };

      if (single_map.hasOwnProperty(ch)) {
        result.push(single_map[ch]);
      } else {
        result.push(ch);
      }
      i += 1;
    }

    return result.join('');
  }

  function cyrillicToLatin(text) {
    if (!text) return '';
    var len = text.length;
    var result = [];
    var i = 0;

    while (i < len) {
      var ch = text[i];
      var next_ch = i + 1 < len ? text[i + 1] : '';

      // 1. Non-digraph compound: сҳ -> s'h
      if ((ch === 'с' || ch === 'С') && (next_ch === 'ҳ' || next_ch === 'Ҳ')) {
        if (ch === 'С' && next_ch === 'Ҳ') result.push("S'H");
        else if (ch === 'С') result.push("S'h");
        else if (next_ch === 'Ҳ') result.push("s'H");
        else result.push("s'h");
        i += 2;
        continue;
      }

      // 2. Contextual 'Е' / 'е'
      if (ch === 'Е' || ch === 'е') {
        var prev_ch = i > 0 ? text[i - 1] : '';
        var is_at_start = i === 0 || WORD_BOUNDARY_CHARS.has(prev_ch);
        var is_after_vowel = CYRILLIC_VOWELS.has(prev_ch) || LATIN_VOWELS.has(prev_ch);

        if (is_at_start || is_after_vowel) {
          if (ch === 'Е') {
            result.push(isUpperCaseContext(text, i) ? 'YE' : 'Ye');
          } else {
            result.push('ye');
          }
        } else {
          result.push(ch === 'Е' ? 'E' : 'e');
        }
        i += 1;
        continue;
      }

      // 3. 'Э' / 'э'
      if (ch === 'Э') { result.push('E'); i += 1; continue; }
      if (ch === 'э') { result.push('e'); i += 1; continue; }

      // 4. Multi-letter representations: Ш, Ч, Ё, Ю, Я, Ц, Ў, Ғ, Щ
      var upper_ctx = isUpperCaseContext(text, i);

      var cyr_multi_map = {
        'Ш': upper_ctx ? 'SH' : 'Sh',
        'ш': 'sh',
        'Ч': upper_ctx ? 'CH' : 'Ch',
        'ч': 'ch',
        'Ё': upper_ctx ? 'YO' : 'Yo',
        'ё': 'yo',
        'Ю': upper_ctx ? 'YU' : 'Yu',
        'ю': 'yu',
        'Я': upper_ctx ? 'YA' : 'Ya',
        'я': 'ya',
        'Ц': upper_ctx ? 'TS' : 'Ts',
        'ц': 'ts',
        'Ў': "O'",
        'ў': "o'",
        'Ғ': "G'",
        'ғ': "g'",
        'Щ': upper_ctx ? 'SHCH' : 'Shch',
        'щ': 'shch',
        'Ъ': "'",
        'ъ': "'",
        'Ь': '',
        'ь': '',
      };

      if (cyr_multi_map.hasOwnProperty(ch)) {
        result.push(cyr_multi_map[ch]);
        i += 1;
        continue;
      }

      // 5. Single Cyrillic letters
      var single_cyr_map = {
        'А': 'A', '\u0430': 'a',
        'Б': 'B', 'б': 'b',
        'В': 'V', 'в': 'v',
        'Г': 'G', 'г': 'g',
        'Д': 'D', 'д': 'd',
        'Ж': 'J', 'ж': 'j',
        'З': 'Z', '\u0437': 'z',
        'И': 'I', 'и': 'i',
        'Й': 'Y', 'й': 'y',
        'К': 'K', '\u043A': 'k',
        'Қ': 'Q', '\u049B': 'q',
        'Л': 'L', 'л': 'l',
        'М': 'M', 'м': 'm',
        'Н': 'N', '\u043D': 'n',
        'О': 'O', '\u043E': 'o',
        'П': 'P', 'п': 'p',
        'Р': 'R', '\u0440': 'r',
        'С': 'S', '\u0441': 's',
        'Т': 'T', '\u0442': 't',
        'У': 'U', '\u0443': 'u',
        'Ф': 'F', 'ф': 'f',
        'Х': 'X', '\u0445': 'x',
        'Ҳ': 'H', '\u04B3': 'h',
        'Ы': 'I', 'ы': 'i',
      };

      if (single_cyr_map.hasOwnProperty(ch)) {
        result.push(single_cyr_map[ch]);
      } else {
        result.push(ch);
      }
      i += 1;
    }

    return result.join('');
  }

  function transliterate(text, direction) {
    direction = direction || 'latin-to-cyrillic';
    if (direction === 'latin-to-cyrillic') {
      return latinToCyrillic(text);
    } else if (direction === 'cyrillic-to-latin') {
      return cyrillicToLatin(text);
    } else {
      throw new Error('Unknown transliteration direction: ' + direction);
    }
  }

  return {
    normalizeApostrophes: normalizeApostrophes,
    latinToCyrillic: latinToCyrillic,
    cyrillicToLatin: cyrillicToLatin,
    transliterate: transliterate
  };
}));
